Loopback HTTP server automatically shuts down following authorization.

WilliamDenniss opened the pull request:
- Automatically closes the loopback HTTP server once an authorization response is received.
- Now cancels the previous request if OIDRedirectHTTPHandler::startHTTPListener is called more than once.
- Removed dead code in OIDLoopbackHTTPServer related to the default behavior when no delegate set to handle connections.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/openid/appauth-ios/53)
<!-- Reviewable:end -->

StevenEWright commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 311 at r1](https://reviewable.io:443/reviews/openid/appauth-ios/53#-KVh5fZiThUVoZBAy2Xu:-KVh5fZjo1deTDxzll8I:bj41l30) ([raw file](https://github.com/openid/appauth-ios/blob/3e2feb1fb2c75c98d9c1984d80885aad6eec9acb/Example-Mac/Source/AppAuthExampleViewController.m#L311)):*
> ```objc
> 
>       // The loopback HTTP listener is no longer needed, stops it.
>       [_redirectHTTPHandler stopHTTPListener];
> ```

Did we remove these because they are no longer needed thanks to the changes in _this_ change?

---

*[Source/macOS/OIDRedirectHTTPHandler.h, line 66 at r1](https://reviewable.io:443/reviews/openid/appauth-ios/53#-KVh58izUUOGLyhkWi2L:-KVh58izUUOGLyhkWi2M:b-jb9l0f) ([raw file](https://github.com/openid/appauth-ios/blob/0a16a44b2592ca1489f11ce71cf2085071763459/Source/macOS/OIDRedirectHTTPHandler.h#L66)):*
> ```objc
>  *      be called except when abandoning the authorization request.
>  */
> - (void)cancelHTTPListener;
> ```

We should probably be reluctant to make public API changes, generally... especially purely aesthetic changes. But it's your call. And I realize this method doesn't really fall into the category of changes I'm specifically talking about... so I have no very strong opinion here. Just thought I should mention it.

---

*[Source/macOS/OIDRedirectHTTPHandler.m, line 74 at r1](https://reviewable.io:443/reviews/openid/appauth-ios/53#-KVh5ykq40nC28RTMM4n:-KVh5ykq40nC28RTMM4o:b-1jyriu) ([raw file](https://github.com/openid/appauth-ios/blob/0a16a44b2592ca1489f11ce71cf2085071763459/Source/macOS/OIDRedirectHTTPHandler.m#L74)):*
> ```objc
> }
> 
> - (void)stopHTTPListener {
> ```

Documentation?

---

*[Source/macOS/OIDRedirectHTTPHandler.m, line 80 at r1](https://reviewable.io:443/reviews/openid/appauth-ios/53#-KVh7Qkbq6QvkLrIekdQ:-KVh7Qkbq6QvkLrIekdR:b-niv13p) ([raw file](https://github.com/openid/appauth-ios/blob/0a16a44b2592ca1489f11ce71cf2085071763459/Source/macOS/OIDRedirectHTTPHandler.m#L80)):*
> ```objc
> }
> 
> - (void)HTTPConnection:(HTTPConnection *)conn didReceiveRequest:(HTTPServerRequest *)mess {
> ```

Would it make any difference to potential flows below if you return fast if (!_currentAuthorizationFlow) ?

I'm not suggesting it. Just curious.

---

*[Source/macOS/OIDRedirectHTTPHandler.m, line 87 at r1](https://reviewable.io:443/reviews/openid/appauth-ios/53#-KVh6lj4eCO2xLduhslw:-KVh6lj4eCO2xLduhslx:b-llfmwj) ([raw file](https://github.com/openid/appauth-ios/blob/0a16a44b2592ca1489f11ce71cf2085071763459/Source/macOS/OIDRedirectHTTPHandler.m#L87)):*
> ```objc
>   // Stops listening to further requests after the first valid authorization response.
>   if (handled) {
>     _currentAuthorizationFlow = nil;
> ```

Doesn't this happen every time stopHTTPListener is called?

Shouldn't it happen every time?

---

*[Source/macOS/OIDRedirectHTTPHandler.m, line 95 at r1](https://reviewable.io:443/reviews/openid/appauth-ios/53#-KVh6yHfnhaC5ig-hSRd:-KVh6yHfnhaC5ig-hSRe:btj1zmy) ([raw file](https://github.com/openid/appauth-ios/blob/0a16a44b2592ca1489f11ce71cf2085071763459/Source/macOS/OIDRedirectHTTPHandler.m#L95)):*
> ```objc
>   NSInteger httpResponseCode;
>   if (handled) {
>     bodyText = @"<html><body>Authorization complete.<br> Return to the app.</body></html>";
> ```

It may be clearer to merge your two if (handled) { ... } cases. The only thing separating them are some symbol declarations. I understand the separation in the context of this change... but not sure future onlookers will. Up to you.

---

*[Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.h, line 95 at r1](https://reviewable.io:443/reviews/openid/appauth-ios/53#-KVh7fLijMSP00FTBKR5:-KVh7fLijMSP00FTBKR6:b-qt551i) ([raw file](https://github.com/openid/appauth-ios/blob/3e2feb1fb2c75c98d9c1984d80885aad6eec9acb/Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.h#L95)):*
> ```objc
> - (void)setConnectionClass:(Class)value;
> 
> - (NSURL *)documentRoot;
> ```

Why are we changing this code?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openid/appauth-ios/53#-:-KVh86kyWYeD9ae6Qhc_:bnfp4nl)*
<!-- Sent from Reviewable.io -->

WilliamDenniss commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[Example-Mac/Source/AppAuthExampleViewController.m, line 311 at r1](https://reviewable.io:443/reviews/openid/appauth-ios/53#-KVh5fZiThUVoZBAy2Xu:-KVhB458Glwev6obDjhf:b-j7tqcl) ([raw file](https://github.com/openid/appauth-ios/blob/3e2feb1fb2c75c98d9c1984d80885aad6eec9acb/Example-Mac/Source/AppAuthExampleViewController.m#L311)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> Did we remove these because they are no longer needed thanks to the changes in _this_ change?
</details>
Yes. It's no longer needed to stop the listener (in fact, that method is gone – replaced with `cancelHTTPListener` if you really want to cancel it out of band).

You can still nil it if you like but it's a very lightweight object once the auth is complete (basically the size of a NSString), so for simplicity (avoiding the need for a strongSelf), I'm allowing it to get deallocated with the ViewController.

---

*[Source/macOS/OIDRedirectHTTPHandler.h, line 66 at r1](https://reviewable.io:443/reviews/openid/appauth-ios/53#-KVh58izUUOGLyhkWi2L:-KVhBWTtL3u16uPBn1e3:b-xm31hs) ([raw file](https://github.com/openid/appauth-ios/blob/0a16a44b2592ca1489f11ce71cf2085071763459/Source/macOS/OIDRedirectHTTPHandler.h#L66)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> We should probably be reluctant to make public API changes, generally... especially purely aesthetic changes. But it's your call. And I realize this method doesn't really fall into the category of changes I'm specifically talking about... so I have no very strong opinion here. Just thought I should mention it.
</details>
It's a fair point, and if it that method was more than a few days old I would have. Technically with semantic versioning since we're a 0.x product, API changes are OK without major version bump (but even then I wouldn't do it hastily).

---

*[Source/macOS/OIDRedirectHTTPHandler.m, line 74 at r1](https://reviewable.io:443/reviews/openid/appauth-ios/53#-KVh5ykq40nC28RTMM4n:-KVhGSojLU-_3cIYzmIF:b-896fix) ([raw file](https://github.com/openid/appauth-ios/blob/0a16a44b2592ca1489f11ce71cf2085071763459/Source/macOS/OIDRedirectHTTPHandler.m#L74)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> Documentation?
</details>
Done.

---

*[Source/macOS/OIDRedirectHTTPHandler.m, line 80 at r1](https://reviewable.io:443/reviews/openid/appauth-ios/53#-KVh7Qkbq6QvkLrIekdQ:-KVhGUCMKaq3FgmpSpKs:bhyi0gz) ([raw file](https://github.com/openid/appauth-ios/blob/0a16a44b2592ca1489f11ce71cf2085071763459/Source/macOS/OIDRedirectHTTPHandler.m#L80)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> Would it make any difference to potential flows below if you return fast if (!_currentAuthorizationFlow) ?
> 
> I'm not suggesting it. Just curious.
</details>
Potentially. If this request was from the user it would indicate a developer error (as they started the authorization without setting the currentAuthorizationFlow object), though it's possible something could hit the server unrelated to the authorization flow after the port was opened, and before it was kicked off.

---

*[Source/macOS/OIDRedirectHTTPHandler.m, line 87 at r1](https://reviewable.io:443/reviews/openid/appauth-ios/53#-KVh6lj4eCO2xLduhslw:-KVhByZoucHixvYQLXDv:bh50mfm) ([raw file](https://github.com/openid/appauth-ios/blob/0a16a44b2592ca1489f11ce71cf2085071763459/Source/macOS/OIDRedirectHTTPHandler.m#L87)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> Doesn't this happen every time stopHTTPListener is called?
> 
> Shouldn't it happen every time?
</details>
`stopHTTPListener` only stops the listener. I thought about adding this behavior there, but then decided to make an explicit public `cancelHTTPListener`. _currentAuthorizationFlow shouldn't ever be set to nil without a success/fail message being sent. `cancelHTTPListener` sends the fail, this block of code is the success version of the same.

---

*[Source/macOS/OIDRedirectHTTPHandler.m, line 95 at r1](https://reviewable.io:443/reviews/openid/appauth-ios/53#-KVh6yHfnhaC5ig-hSRd:-KVhCgYk9pgv6cbT4B1Q:by8yqja) ([raw file](https://github.com/openid/appauth-ios/blob/0a16a44b2592ca1489f11ce71cf2085071763459/Source/macOS/OIDRedirectHTTPHandler.m#L95)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> It may be clearer to merge your two if (handled) { ... } cases. The only thing separating them are some symbol declarations. I understand the separation in the context of this change... but not sure future onlookers will. Up to you.
</details>
I see your point. Was trying to keep the functionality separated between the internal API calls (closing the listener, sending success message), and sending the HTTP server response.

I refactored a little, does it help?

---

*[Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.h, line 95 at r1](https://reviewable.io:443/reviews/openid/appauth-ios/53#-KVh7fLijMSP00FTBKR5:-KVhDadWoudrEULrQtAE:b-mpkt8o) ([raw file](https://github.com/openid/appauth-ios/blob/3e2feb1fb2c75c98d9c1984d80885aad6eec9acb/Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.h#L95)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> Why are we changing this code?
</details>
Unrelated change that I noticed when looking at the HTTP handling. I've moved it to a separate commit, but still in this same PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openid/appauth-ios/53)*
<!-- Sent from Reviewable.io -->

StevenEWright commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[Source/macOS/OIDRedirectHTTPHandler.m, line 80 at r1](https://reviewable.io:443/reviews/openid/appauth-ios/53#-KVh7Qkbq6QvkLrIekdQ:-KVhLgHwQIGgwUemnHYT:bqfh4em) ([raw file](https://github.com/openid/appauth-ios/blob/0a16a44b2592ca1489f11ce71cf2085071763459/Source/macOS/OIDRedirectHTTPHandler.m#L80)):*
<details><summary><i>Previously, WilliamDenniss (William Denniss) wrote…</i></summary>
> Potentially. If this request was from the user it would indicate a developer error (as they started the authorization without setting the currentAuthorizationFlow object), though it's possible something could hit the server unrelated to the authorization flow after the port was opened, and before it was kicked off.
</details>
As long as you've considered it :) 

---

*[Source/macOS/OIDRedirectHTTPHandler.m, line 95 at r1](https://reviewable.io:443/reviews/openid/appauth-ios/53#-KVh6yHfnhaC5ig-hSRd:-KVhM4jV7U0qJbqeCa2i:b-2v4vsf) ([raw file](https://github.com/openid/appauth-ios/blob/0a16a44b2592ca1489f11ce71cf2085071763459/Source/macOS/OIDRedirectHTTPHandler.m#L95)):*
<details><summary><i>Previously, WilliamDenniss (William Denniss) wrote…</i></summary>
> I see your point. Was trying to keep the functionality separated between the internal API calls (closing the listener, sending success message), and sending the HTTP server response.
> 
> I refactored a little, does it help?
</details>
http://gph.is/NY2QTR

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openid/appauth-ios/53#-:-KVhMDPv_2kmCSHm32wU:bnfp4nl)*
<!-- Sent from Reviewable.io -->

WilliamDenniss commented:
Review status: 3 of 5 files reviewed at latest revision, all discussions resolved.

---

*[Source/macOS/OIDRedirectHTTPHandler.m, line 80 at r1](https://reviewable.io:443/reviews/openid/appauth-ios/53#-KVh7Qkbq6QvkLrIekdQ:-KVi0ULgBt4Yl4WDOCaq:b-jz60o0) ([raw file](https://github.com/openid/appauth-ios/blob/0a16a44b2592ca1489f11ce71cf2085071763459/Source/macOS/OIDRedirectHTTPHandler.m#L80)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> As long as you've considered it :) 
</details>
In the latest commit, I've added a specific HTML error message for this case to help the developer solve the misconfiguration issue of not setting a `currentAuthorizationFlow`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openid/appauth-ios/53)*
<!-- Sent from Reviewable.io -->
